relay-server/src/sql/mod.rs

```diff
@@ -0,0 +1,100 @@
+use futures::stream::{Stream, StreamExt};
+use sqlx::query::Query;
+use sqlx::sqlite::SqliteArguments;
+use sqlx::{Pool, QueryBuilder, Sqlite};
+
+use crate::actors::project_buffer::QueueKey;
+use crate::statsd::RelayCounters;
+
+/// SQLite allocates space to hold all host parameters between 1 and the largest host parameter number used.
+///
+/// To prevent excessive memory allocations, the maximum value of a host parameter number is SQLITE_MAX_VARIABLE_NUMBER,
+/// which defaults to 999 for SQLite versions prior to 3.32.0 (2020-05-22) or 32766 for SQLite versions after 3.32.0.
+///
+/// Keep it on the lower side for now.
+const SQLITE_LIMIT_VARIABLE_NUMBER: usize = 999;
+
+/// Creates a DELETE query binding to the provided [`QueueKey`] which returns the envelopes and
+/// timestamp.
+///
+/// The query will perform the delete once executed returning deleted envelope and timestamp when
+/// the envelope was received. This will create a prepared statement which is cached and re-used.
+pub fn delete_and_fetch<'a>(
+    key: QueueKey,
+    batch_size: u32,
+) -> Query<'a, Sqlite, SqliteArguments<'a>> {
+    sqlx::query(
+        "DELETE FROM
+            envelopes
+         WHERE id IN (SELECT id FROM envelopes WHERE own_key = ? AND sampling_key = ? LIMIT ?)
+         RETURNING
+            envelope, received_at",
+    )
+    .bind(key.own_key.to_string())
+    .bind(key.sampling_key.to_string())
+    .bind(batch_size)
+}
+
+/// Creates a DELETE query, which silently removes the data from the database.
+pub fn delete<'a>(key: QueueKey) -> Query<'a, Sqlite, SqliteArguments<'a>> {
+    sqlx::query("DELETE FROM envelopes where own_key = ? AND sampling_key = ?")
+        .bind(key.own_key.to_string())
+        .bind(key.sampling_key.to_string())
+}
+
+/// Creates a query which fetches the current page count and page size.
+///
+/// This info used to calculate the current allocated database size.
+pub fn current_size<'a>() -> Query<'a, Sqlite, SqliteArguments<'a>> {
+    sqlx::query("pragma page_count; pragma page_size;")
+}
+
+/// Descibes the chunk item which is handled by insert statement.
+type ChunkItem = (QueueKey, Vec<u8>, i64);
+
+/// Creates an INSERT query for the chunk of provided data.
+pub fn insert_with_builder<'a>(
```

Comment:
AFAICT this is not used outside of this module.

```suggestion
fn insert_with_builder<'a>(
```